### ISSUE TITLE
An incorrect query result by missing a row

### ISSUE THREAD

**Issue by sayJason** [pingcap/tidb#41877]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```SQL
CREATE TABLE t0(c0 INTEGER ZEROFILL NOT NULL );
INSERT INTO t0 VALUES (1);
SELECT * FROM t0 WHERE c0 < BIN(1265859629); -- actual:{}, expected: {1}
```
### 2. What did you expect to see? (Required)
The SELECT statement returns the value 1.
### 3. What did you see instead (Required)
The SELECT statement returns an empty result.
### 4. What is your TiDB version? (Required)

Release Version: v6.6.0
Edition: Community
Git Commit Hash: f4ca0821fb96a2bdd37d2fb97eb26c07fc58d4e4
Git Branch: heads/refs/tags/v6.6.0
UTC Build Time: 2023-02-17 14:49:02
GoVersion: go1.19.5
Race Enabled: false
TiKV Min Version: 6.2.0-alpha
Check Table Before Drop: false
Store: tikv

**Comment by yibin87** [comment]
Not reproduced by v8.0.0 version:
Release Version: v8.0.0
Edition: Community
Git Commit Hash: 8ba1fa452b1ccdbfb85879ea94b9254aabba2916
Git Branch: HEAD
UTC Build Time: 2024-03-28 14:22:04
GoVersion: go1.21.6
Race Enabled: false
Check Table Before Drop: false
Store: tikv

**Comment by yibin87** [comment]
Still exists in v7.5.0, seems fixed but not cherry picked to v7.5.0 and earlier version.

**Comment by yibin87** [comment]
Close it now, please fill free to open it, if reproduce it with version newer than v8.0.0.
